examples/nlp/data_parallel_training_with_keras_nlp.py

```diff
@@ -0,0 +1,211 @@
+Synchronicity keeps the model convergence behavior identical to what you would see for 
+single-device training.
+
+Specifically, this guide teaches you how to use the `tf.distribute` API to train Keras
+models on multiple GPUs, with minimal changes to your code, in the following two setups:
+
+- On multiple GPUs (typically 2 to 8) installed on a single machine (single host,
+multi-device training). This is the most common setup for researchers and small-scale
+industry workflows.
+- On a cluster of many machines, each hosting one or multiple GPUs (multi-worker
+distributed training). This is a good setup for large-scale industry workflows, e.g.
+training high-resolution image classification models on tens of millions of images using
+20-100 GPUs.
+
+## Setup
+
+The tutorial relies on KerasNLP 0.5.2. Additionally, we need
+at least TensorFlow 2.11 in order to use AdamW with mixed precision.
+"""
+
+"""shell
+pip install keras-nlp==0.5.2 -q
+pip install -U tensorflow -q
+"""
+
+"""
+## Imports
+"""
+
+import os
+import tensorflow as tf
+from tensorflow import keras
+import keras_nlp
+
+"""
+A simple thing we can do right off the bat is to enable
+[mixed precision](https://keras.io/api/mixed_precision/), which will speed up training by
+running most of our computations with 16 bit (instead of 32 bit) floating point numbers.
+"""
+
+policy = keras.mixed_precision.Policy("mixed_float16")
+keras.mixed_precision.set_global_policy(policy)
+
+
+PRETRAINING_BATCH_SIZE = 128
+EPOCHS = 5
+
+"""
+First, we need to download and preprocess the wikitext-2 dataset. This dataset will be 
+used for pretraining the BERT model. We will filter out short lines to ensure that the 
+data has enough context for training.
+"""
+
+keras.utils.get_file(
+    origin="https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-2-v1.zip",
+    extract=True,
+)
+wiki_dir = os.path.expanduser("~/.keras/datasets/wikitext-2/")
+
+# Load wikitext-103 and filter out short lines.
+wiki_train_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.train.tokens")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+)
+wiki_val_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.valid.token")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+)
+wiki_test_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.test.tokens")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+)
+
+"""
+In the above code, we download the wikitext-2 dataset and extract it. Then, we define 
+three datasets: wiki_train_ds, wiki_val_ds, and wiki_test_ds. These datasets are 
+filtered to remove short lines and are batched for efficient training.
+
+Let's define a function for decaying the learning rate. You can define any decay function you need.
+We also define a callback for printing the learning rate at the end of each epoch.
+"""
+
+
+def decay(epoch):
```

Comment:
Any specific reason for using a per-epoch decay compared to a normal `LearningRateSchedule`? The latter approach is more commonly used IMO

Reply:
@chenmoneygithub  On a second thought, I think we should remove the learning rate decay. Focusing only on the data parallel training demonstration.
Other keras tutorials better demonstrate using LR decay to improve accuracy.

Reply:
Let's throw a simple learning rate schedule:
```
lr = tf.keras.optimizers.schedules.PolynomialDecay(
    5e-5,
    decay_steps={total_training_steps},
    end_learning_rate=0.0,
)
```

It's a common practice to use a decayed learning rate in NLP training/tuning. A callback is a bit over complex, but this should work.